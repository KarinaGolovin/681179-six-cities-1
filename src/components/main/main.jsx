import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {toggleFavorite} from '../../store/actions';
import PlacesList from '../places-list/places-list.jsx';
import CitiesList from '../cities-list/cities-list.jsx';
import Map from '../map/map.jsx';
import {getCityOffers, getCoordinatesByCity, getSelectedCity} from '../../store/reducers/offers/selectors';
import withActiveItem from '../../hocs/with-active-item/with-active-item';
import {withOffersSorting} from '../../hocs/with-offer-sorting/with-offer-sorting';
import SelectForm from '../select-form/select-form.jsx';
import {compose, withProps} from 'recompose';

const PlacesListWrapped = withProps({
  classes: {
    container: `cities__places-list tabs__content`,
  },
  cardProps: {
    classes: {
      container: `cities__place-card`,
      imageWrapper: `cities__image-wrapper`
    }
  }
})(PlacesList);

export const Main = ({
  coordinatesByCity,
  currentPlaces,
  currentCity,
  updateBookmark,
  sortingOptions,
  sortingType,
  onSortTypeChange = () => {},
  onActiveItemChange = () => {},
  activeItem
}) => {
  return (
    <main className="page__main page__main--index">
      <h1 className="visually-hidden">Cities</h1>
      <div className="cities tabs">
        <section className="locations container">
          <CitiesList
            cities={Object.keys(coordinatesByCity)}
            activeItem={currentCity}
          />
        </section>
      </div>
      <div className="cities__places-wrapper">
        <div className="cities__places-container container">
          <section className="cities__places places">
            <h2 className="visually-hidden">Places</h2>
            {currentCity ? (
              <>
                <b className="places__found">{currentPlaces.length} places to stay in {currentCity}</b>
                <SelectForm
                  defaultValue={sortingType}
                  options={sortingOptions}
                  onSortTypeChange={onSortTypeChange}
                />
              </>
            ) : null}
            <PlacesListWrapped
              offers={currentPlaces}
              onBookmarkClick={updateBookmark}
              onActiveItemChange={onActiveItemChange}
            />
          </section>
          <div className="cities__right-section">
            {coordinatesByCity[currentCity] ? (
              <Map
                mapClass={`cities__map`}
                cityCoords={coordinatesByCity[currentCity]}
                placesList={currentPlaces}
                activePlaceId={activeItem ? activeItem.id : null}
              />
            ) : null}
          </div>
        </div>
      </div>
    </main>
  );
};

Main.propTypes = {
  mapZoom: PropTypes.number,
  currentCity: PropTypes.string,
  coordinatesByCity: PropTypes.object,
  currentPlaces: PropTypes.arrayOf(PropTypes.object).isRequired,
  updateBookmark: PropTypes.func,
  onSortTypeChange: PropTypes.func,
  onActiveItemChange: PropTypes.func,
  activeItem: PropTypes.object,
  sortingType: PropTypes.string,
  sortingOptions: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string,
    label: PropTypes.string,
  }))
};

const mapStateToProps = (state, {selectedCityName}) => {
  const currentCity = getSelectedCity(state, selectedCityName);

  return {
    currentCity,
    currentPlaces: getCityOffers(state, currentCity),
    coordinatesByCity: getCoordinatesByCity(state),
  };
};

const mapDispatchToProps = {
  updateBookmark: toggleFavorite
};

export default compose(
    connect(
        mapStateToProps,
        mapDispatchToProps
    ),
    withOffersSorting,
    withActiveItem
)(Main);
