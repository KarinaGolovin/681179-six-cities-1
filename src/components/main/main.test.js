import React from 'react';
import renderer from 'react-test-renderer';
import {Main} from './main.jsx';

const mock = [
  {
    title: `Test`,
    type: `Test`,
    price: 0,
    previewImage: ``,
    link: `#`,
    rating: 0,
    isPremium: true,
    id: `001`,
    city: `Test`,
  },
  {
    title: `Test`,
    type: `Test`,
    price: 0,
    previewImage: ``,
    link: `#`,
    rating: 0,
    isPremium: true,
    id: `002`,
    city: `Amsterdam`,
  },
  {
    title: `Test`,
    type: `Test`,
    price: 0,
    previewImage: ``,
    link: `#`,
    rating: 0,
    isPremium: true,
    id: `003`,
    city: `Amsterdam`,
  },
  {
    title: `Test`,
    type: `Test`,
    price: 0,
    previewImage: ``,
    link: `#`,
    rating: 0,
    isPremium: true,
    id: `004`,
    city: `Amsterdam`,
  },
];

it(`Main correctly renders`, () => {
  const component = renderer.create(
      <Main
        currentCity={`Test`}
        currentPlaces={mock}
        setNewCity={() => { }}
        coordinatesByCity={{
          Test: [0, 0]
        }}
      />
  ).toJSON();

  expect(component).toMatchSnapshot();
});
