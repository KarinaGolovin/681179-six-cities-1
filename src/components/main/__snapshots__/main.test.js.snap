// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main correctly renders 1`] = `
<main
  className="page__main page__main--index"
>
  <h1
    className="visually-hidden"
  >
    Cities
  </h1>
  <div
    className="cities tabs"
  >
    <section
      className="locations container"
    >
      <ul
        className="locations__list tabs__list"
      >
        <li
          className="locations__item"
        >
          <a
            className="locations__item-link tabs__item tabs__item--active"
            href="/city/Test"
            onClick={[Function]}
          >
            <span>
              Test
            </span>
          </a>
        </li>
      </ul>
    </section>
  </div>
  <div
    className="cities__places-wrapper"
  >
    <div
      className="cities__places-container container"
    >
      <section
        className="cities__places places"
      >
        <h2
          className="visually-hidden"
        >
          Places
        </h2>
        <b
          className="places__found"
        >
          4
           places to stay in 
          Test
        </b>
        <form
          action="#"
          className="places__sorting"
          method="get"
        >
          <span
            className="places__sorting-caption"
          >
            Sort by
          </span>
           
          <span
            className="places__sorting-type"
            tabIndex="0"
          >
            Popular
            <svg
              className="places__sorting-arrow"
              height="4"
              width="7"
            >
              <use
                xlinkHref="#icon-arrow-select"
              />
            </svg>
          </span>
          <select
            className="places__sorting-type"
            id="places-sorting"
          >
            <option
              className="places__option"
              defaultValue=""
              value="popular"
            >
              Popular
            </option>
            <option
              className="places__option"
              value="to-high"
            >
              Price: low to high
            </option>
            <option
              className="places__option"
              value="to-low"
            >
              Price: high to low
            </option>
            <option
              className="places__option"
              value="top-rated"
            >
              Top rated first
            </option>
          </select>
        </form>
        <div
          className="cities__places-list places__list tabs__content"
        >
          <article
            className="place-card cities__place-card"
            id={1}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="place-card__image-wrapper cities__image-wrapper"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height={200}
                  width={260}
                />
              </a>
            </div>
            <div
              className="place-card__info "
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    0
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button  button"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height="19"
                    width="18"
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    To bookmarks
                  </span>
                </button>
              </div>
              <div
                className="rating place-card__rating"
              >
                <div
                  className="rating__stars place-card__stars"
                >
                  <span
                    style={
                      Object {
                        "width": "0%",
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                <a
                  href="#"
                  onClick={[Function]}
                >
                  Test
                </a>
              </h2>
              <p
                className="place-card__type"
              >
                Test
              </p>
            </div>
          </article>
          <article
            className="place-card cities__place-card"
            id={2}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="place-card__image-wrapper cities__image-wrapper"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height={200}
                  width={260}
                />
              </a>
            </div>
            <div
              className="place-card__info "
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    0
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button  button"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height="19"
                    width="18"
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    To bookmarks
                  </span>
                </button>
              </div>
              <div
                className="rating place-card__rating"
              >
                <div
                  className="rating__stars place-card__stars"
                >
                  <span
                    style={
                      Object {
                        "width": "0%",
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                <a
                  href="#"
                  onClick={[Function]}
                >
                  Test
                </a>
              </h2>
              <p
                className="place-card__type"
              >
                Test
              </p>
            </div>
          </article>
          <article
            className="place-card cities__place-card"
            id={3}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="place-card__image-wrapper cities__image-wrapper"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height={200}
                  width={260}
                />
              </a>
            </div>
            <div
              className="place-card__info "
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    0
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button  button"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height="19"
                    width="18"
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    To bookmarks
                  </span>
                </button>
              </div>
              <div
                className="rating place-card__rating"
              >
                <div
                  className="rating__stars place-card__stars"
                >
                  <span
                    style={
                      Object {
                        "width": "0%",
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                <a
                  href="#"
                  onClick={[Function]}
                >
                  Test
                </a>
              </h2>
              <p
                className="place-card__type"
              >
                Test
              </p>
            </div>
          </article>
          <article
            className="place-card cities__place-card"
            id={4}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="place-card__image-wrapper cities__image-wrapper"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height={200}
                  width={260}
                />
              </a>
            </div>
            <div
              className="place-card__info "
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    0
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button  button"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height="19"
                    width="18"
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    To bookmarks
                  </span>
                </button>
              </div>
              <div
                className="rating place-card__rating"
              >
                <div
                  className="rating__stars place-card__stars"
                >
                  <span
                    style={
                      Object {
                        "width": "0%",
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                <a
                  href="#"
                  onClick={[Function]}
                >
                  Test
                </a>
              </h2>
              <p
                className="place-card__type"
              >
                Test
              </p>
            </div>
          </article>
        </div>
      </section>
      <div
        className="cities__right-section"
      >
        <section
          className="cities__map map"
          id="map"
        />
      </div>
    </div>
  </div>
</main>
`;
